rfcs/hse/0004-kvdb-kvs-introspection-apis.md

```diff
@@ -0,0 +1,154 @@
+  HSE_MCLASS_STAGING,
+};
+
+struct hse_kvdb_storage_info {
+  uint64_t total_bytes;     /**< total space in the file-system containing this kvdb */
+  uint64_t available_bytes; /**< available space in the file-system containing this kvdb */
+  uint64_t allocated_bytes; /**< allocated storage space for a kvdb */
+  uint64_t used_bytes;      /**< used storage space for a kvdb */
+  char     path[PATH_MAX];  /**< path to media class */
+};
+
+/* Populates the above storage info struct with given information.
+ *
+ * Con: Since we only track data owned by each media class, we drop the storage
+ * statistics about the kvdb.meta file, but I don't see that as being super
+ * important.
+ */
+hse_err_t
+hse_kvdb_storage_info_get(const struct hse_kvdb *kvdb, struct hse_kvdb_storage_info *info);
+```
+
+#### REST APIs
+
+One issue we have currently that is worth solving in this RFC is that our REST
+API assumes there is only one KVDB. We seem to have settled on the idea we want
+to support more than one KVDB in the future. Now is the perfect time to retrofit
+the REST API.
+
+One issue with name-less KVDBs is that file paths look like parts of a REST
+path, which is not good for easy usage. Typically in a well-planned REST route
+tree, you would have something like `/kvdb/<specific kvdb>`, where `/kvdb` is
+the parent route of all KVDBs. In our current, name-less KVDB situation this
+would look like `/kvdb/path/to/kvdb` where `path/to/kvdb` is the KVDB home. This
+makes it impossible to parse REST paths correctly. What we can do instead is the
+following:
+
+- Add a new KVDB rparam: `alias`
+  - The value of `alias` would default to `0` for the first KVDB opened, `1` for
+    the next KVDB opened and so on. In the event, KVDB 0 is closed, and KVDB 2
+    is opened, KVDB 2 would take over the alias 0. Alex has suggested using an
+    always incrementing alias.
+  - The rparam could also be called `socket.alias`, but I have a feeling the
+    alias could come in handy in other sections of our code, and it doesn't
+    hamstring us.
+
+The following REST APIs will be added:
+
+- `/kvdb`
+  - `GET` - Return a JSON body of all currently opened KVDBs
+- `/kvdb/<alias>/home`
+  - `GET` - Return a JSON body of the KVDB home
+- `/kvdb/<alias>/storage/info`
+  - `GET` - Return a JSON body of the storage info struct
+- `/kvdb/<alias>/params`
+  - `GET` - Return a JSON body of all the KVDB parameters
+- `/kvdb/<alias>/params/<param>`
+  - `GET` - Return a JSON body of just the parameter that was asked for
+  - Requirement: create and runtime params must have unique names
+- `/kvdb/<alias>/kvs`
+  - `GET` - Return a JSON body of all the KVS names in a KVDB
+- `/kvdb/<alias>/kvs/<name>/params`
+  - `GET` - Return a JSON body of all the KVS parameters for specified KVS
+- `/kvdb/<alias>/kvs/<name>/params/<param>`
+  - `GET` - Return a JSON body of just the parameter that was asked for
+  - Requirement: create and runtime params must have unique names
```

Comment:
Since there wasn't an equivalent C API, I didn't feel the need to add a REST API. It's doable, but I am also not entirely sure I see the value.

Reply:
Global parameters are treated in our API, our docs and internally as equals to KVDB and KVS parameters.  To me it would be illogical to leave them out here.  Some global parameters can have great impact on system behavior (e.g., `c0kvs_cheap_sz`).  Others are useful as writable for debug (e.g., `logging.level`).